eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala

```diff
@@ -67,54 +69,93 @@ class PaymentInitiator(nodeParams: NodeParams, router: ActorRef, relayer: ActorR
     case r: SendTrampolinePaymentRequest =>
       val paymentId = UUID.randomUUID()
       sender ! paymentId
-      if (!r.paymentRequest.features.allowTrampoline && r.paymentRequest.amount.isEmpty) {
-        sender ! PaymentFailed(paymentId, r.paymentRequest.paymentHash, LocalFailure(new IllegalArgumentException("cannot pay a 0-value invoice via trampoline-to-legacy (trampoline may steal funds)")) :: Nil)
-      } else {
-        val paymentCfg = SendPaymentConfig(paymentId, paymentId, None, r.paymentRequest.paymentHash, r.trampolineNodeId, Upstream.Local(paymentId), Some(r.paymentRequest), storeInDb = true, publishEvent = true, Some(r))
-        val finalPayload = if (r.paymentRequest.features.allowMultiPart) {
-          Onion.createMultiPartPayload(r.finalAmount, r.finalAmount, r.finalExpiry(nodeParams.currentBlockHeight), r.paymentRequest.paymentSecret.get)
-        } else {
-          Onion.createSinglePartPayload(r.finalAmount, r.finalExpiry(nodeParams.currentBlockHeight), r.paymentRequest.paymentSecret)
-        }
-        val trampolineRoute = Seq(
-          NodeHop(nodeParams.nodeId, r.trampolineNodeId, nodeParams.expiryDeltaBlocks, 0 msat),
-          NodeHop(r.trampolineNodeId, r.paymentRequest.nodeId, r.trampolineExpiryDelta, r.trampolineFees) // for now we only use a single trampoline hop
-        )
-        // We assume that the trampoline node supports multi-part payments (it should).
-        val (trampolineAmount, trampolineExpiry, trampolineOnion) = if (r.paymentRequest.features.allowTrampoline) {
-          OutgoingPacket.buildPacket(Sphinx.TrampolinePacket)(r.paymentRequest.paymentHash, trampolineRoute, finalPayload)
-        } else {
-          OutgoingPacket.buildTrampolineToLegacyPacket(r.paymentRequest, trampolineRoute, finalPayload)
-        }
-        // We generate a random secret for this payment to avoid leaking the invoice secret to the first trampoline node.
-        val trampolineSecret = randomBytes32
-        spawnMultiPartPaymentFsm(paymentCfg) forward SendMultiPartPayment(r.paymentRequest.paymentHash, trampolineSecret, r.trampolineNodeId, trampolineAmount, trampolineExpiry, 1, r.paymentRequest.routingInfo, r.routeParams, Seq(OnionTlv.TrampolineOnion(trampolineOnion.packet)))
+      r.trampolineAttempts match {
+        case Nil =>
+          sender ! PaymentFailed(paymentId, r.paymentRequest.paymentHash, LocalFailure(new IllegalArgumentException("trampoline fees and cltv expiry delta are missing")) :: Nil)
```

Comment:
`LocalFailure` accepts a generic `Throwable` for the failure details, however it would be useful to type this failure, like it was done with `RouteNotFound` or `BalanceTooLow`. This would let the front provide an accurate error message (incl. localization) with potentially a different flow depending on the error.

Ideally `LocalFailure` would force the throwable to be typed. There are several options to do that, but let's not derail this PR. 

In the meantime I think `LocalFailure` should contain a specific object as much as possible.

Reply:
Good idea, I'll create a specific object for that error and will improve the overall situation in the next trampoline PR.

Reply:
Done